### PR TITLE
1.16+ Terrain and Movement support

### PR DESCRIPTION
**Pull request by ReinforceZwei**
related #1090

### CODE REVIEW

**Comment by ReinforceZwei** [comment]
Seems like `MultiBlockChange` packet also got changed but I haven't updated it yet.

**Comment by ReinforceZwei** [comment]
Terrain and movement is fully functional now.

**Comment by Mhowser** [comment]
Excellent job! Thank you for doing this!!

File: MinecraftClient/Protocol/Handlers/Protocol18.cs [hunk 1]
```diff
@@ -413,8 +417,26 @@ internal bool HandlePacket(int packetID, Queue<byte> packetData)
...
+                                {
+                                    if (protocolversion >= MC1162Version)
+                                    {
+                                        for (int i = 0; i < biomesLength; i++)
+                                        {
+                                            // Biomes - 1.16.2 and above
+                                            // Don't use ReadNextVarInt because it cost too much time
+                                            while (true)
+                                            {
+                                                if ((dataTypes.ReadNextByte(packetData) & 0x80) != 128) break;
```
Comment (ORelio) [inline MinecraftClient/Protocol/Handlers/Protocol18.cs:434]: Clever! Seems worth to move into a dedicated method `dataTypes.SkipNextVarInt(packetData);` 👍

File: MinecraftClient/Protocol/Handlers/Protocol18Terrain.cs [hunk 2]
```diff
@@ -79,45 +80,97 @@ public void ProcessChunkColumnData(int chunkX, int chunkZ, ushort chunkMask, ush
 
                         Chunk chunk = new Chunk();
 
-                        if (dataArray.Length > 0)
+                        if (protocolversion >= Protocol18Handler.MC116Version)
```
Comment (ORelio) [inline MinecraftClient/Protocol/Handlers/Protocol18Terrain.cs:83]: I see no changes for 1.16 in https://wiki.vg/Chunk_Format#Chunk_Section_structure ; so I'm not 100% sure what is done here specifically for 1.16. If it's a more optimized chunk parsing routine, it should work also below 1.16 so no need to duplicate the code here, right?

**Review by ORelio — COMMENTED** [review]
Good job in implementing 1.16 chunk data and multi-block changes! I've a question about `ProcessChunkColumnData` before merging:

File: MinecraftClient/Protocol/Handlers/Protocol18Terrain.cs (hunk 2 above)
Comment (ReinforceZwei) [reply MinecraftClient/Protocol/Handlers/Protocol18Terrain.cs:83]: No. The document for chunk format is outdated and doesn't work on 1.16.

Comment (ORelio) [reply MinecraftClient/Protocol/Handlers/Protocol18Terrain.cs]: Thanks for pointing that out, everything makes sense 👍
I adjusted the parsing code to use your approach instead and disable the padding part when parsing 1.15 chunk data.

**Review by ORelio — APPROVED** [review]

**Comment by ORelio** [comment]
Thanks again for your work in adding terrain support for Minecraft 1.16. Now merged 👍

**Comment by yougeshlama** [comment]
awwwwwwwwwwwwww iam so happy finally its implemented :)

**Pull request merged**